src/macros/macrorecorder.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <gtest/gtest_prod.h>
+#include <rigtorp/SPSCQueue.h>
+
+#include <QDebug>
+#include <QObject>
+#include <QtCore>
+
+#include "control/controlpushbutton.h"
+#include "engine/channelhandle.h"
+#include "macros/macro.h"
+#include "recording/recordingmanagerbase.h"
+
+/// The MacroRecorder handles the recording of Macros and the [MacroRecording] controls.
+class MacroRecorder : public RecordingManagerBase {
+    Q_OBJECT
+  public:
+    MacroRecorder();
+
+    void startRecording() override;
+    void stopRecording() override;
+
+    enum Status : uint8_t {
+        /// Nothing is going on
+        Disabled = 0,
+        /// Recording is awaiting assignment to a channel
+        Armed = 1,
+        /// Recording is active
+        Recording = 2,
+    };
+
+    bool isRecordingActive() const override;
+    Status getStatus() const;
+
+    /// Tries to append this cue jump to the currently recorded Macro.
+    /// Returns true if the currently recorded Macro was changed - so only if
+    /// recording is active and the channel handle matches.
+    /// Only called in realtime code.
+    void notifyCueJump(ChannelHandle* channel, double sourceFramePos, double destFramePos);
+
+    size_t getRecordingSize() const;
+    const MacroAction getRecordedAction();
```

Comment:
This is only used for tests, right? Unsure if it even makes sense to only return a single action. Instead, I suggest to make a method that returns a QVector or even a Macro instance from the current data and use that for business logic and testing. And the method should be private IMHO.

Reply:
Yes it is, yes I only need a single Action. Returning a collection would defeat the point. I don't see any harm in having this public. I want to avoid test friends as much as possible to not accidentally test internals.

Reply:
This is a test internal regardless if it's public or not, because it's not used in any production code. If you implement a method that returns a list of macros, you can use it in the production code (i.e. the `stopRecording` method): https://github.com/mixxxdj/mixxx/pull/2989/files#diff-b739851afe957536ff71c1052002ccdfR73-R77
If you only need the first element, your test can do an additional check that the length of the list is correct.

Reply:
Ah, now that is an excellent idea :)

Reply:
8a0ca78